kpatch-build/kpatch-elf.c

```diff
@@ -726,6 +753,119 @@ void kpatch_reindex_elements(struct kpatch_elf *kelf)
+		 *    group_section. This is later used to compare if the reindexed
+		 *    section and group_section matches.
+		 * b. Maintain the secnames, symnames to later update
+		 *    group section members sh_link and sh_info.
+		 */
+		groupsec->groupindex = groupsec->index;
+		data = groupsec->data->d_buf;
+		end = groupsec->data->d_buf + groupsec->data->d_size;
+		data++; /* skip first flag word (e.g. GRP_COMDAT) */
+		gsec = malloc(sizeof(*gsec));
+		if (!gsec)
+			ERROR("malloc");
+		gsec->groupindex = groupsec->index;
+		/* Maintain names of all section link. Use this info after reindexing */
+		gsec->secnames = malloc((groupsec->data->d_size/sizeof(*data)) * sizeof(char **));
+		if (!gsec->secnames)
+			ERROR("malloc");
+		gsec->symnames = malloc((groupsec->data->d_size/sizeof(*data)) * sizeof(char **));
+		if (!gsec->symnames)
+			ERROR("malloc");
+
+		while (data < end) {
+			sec = find_section_by_index(&kelf->sections, *data);
+			if (!sec)
+				ERROR("group section not found");
+			sec->grouped = 1;
+			log_debug("marking section %s (%d) as grouped\n",
+				  sec->name, sec->index);
+
+			gsec->secnames[i] = strdup(sec->name);
+			tsym = find_symbol_by_index(&kelf->symbols, groupsec->sh.sh_info);
+			if (!tsym)
+				ERROR("symbol not found");
+			gsec->symnames[i] = strdup(tsym->name);
+			data++;
+			i++;
+		}
+		list_add_tail(&gsec->list, &kelf->groupsec);
+	}
+}
+
+/*
+ * kpatch-build reindexes the sections and symbols. When these are reindexed,
+ * group section needs reindexing as well.
+ *
+ * In s390x, expolines like __s390_indirect_jump_r1 are the referenced in
+ * .group sections.
+ */
+void kpatch_reindex_group_sections(struct kpatch_elf *kelf)
+{
+	struct section *sec, *subsec, *tsec;
+	struct symbol *tsym;
+	struct group_section *groupsec;
+	unsigned int *data, *end, i = 0;
+	unsigned int *newdata;
+
+	list_for_each_entry(groupsec, &kelf->groupsec, list) {
+		i = 0;
+		list_for_each_entry(sec, &kelf->sections, list) {
+			/* section's groupindex is set, only if it is a group section */
+			if (sec->sh.sh_type != SHT_GROUP)
+				continue;
+			if (sec->groupindex != groupsec->groupindex)
+				continue;
+			data = sec->data->d_buf;
+			end = sec->data->d_buf + sec->data->d_size;
+			newdata = malloc(sec->data->d_size);
+			if (!newdata)
+				ERROR("malloc");
+
+			newdata[0] = GRP_COMDAT;
+			data++;
+			while (data < end) {
+				subsec = find_section_by_name(&kelf->sections, groupsec->secnames[i]);
+				if (!subsec)
+					ERROR("group section");
+
+				tsym = find_symbol_by_name(&kelf->symbols, groupsec->symnames[i]);
+				if (!tsym)
+					ERROR("symbol not found");
+
+				/* Update the new index in the group section */
+				newdata[++i] = subsec->index;
+				tsec = find_section_by_name(&kelf->sections, ".symtab");
+				if (!tsec)
+					ERROR("section not found");
+
+
+				sec->sh.sh_link = tsec->index;
+				sec->sh.sh_info = tsym->index;
+				data++;
+			}
+			sec->data->d_buf = newdata;
```

Comment:
Also, I think this leaks the previous `sec->data->d_buf` ?

Reply:
I  couldnt free it (Elf_Data). Modifying the memory referenced by sec->data->d_buf or Freeing this results in segfault.